addons/partner_autocomplete/models/res_company.py

```diff
@@ -12,3 +20,54 @@ class ResCompany(models.Model):
     def _inverse_partner_gid(self):
         for company in self:
             company.partner_id.partner_gid = company.partner_gid
+
+    def _enrich(self):
+        """
+        This method calls the partner autocomplete service from IAP to enrich the partner related fields
+        of the company.
+        """
+        self.ensure_one()
+
+        company_domain = self._get_company_domain()
+        if not company_domain:
+            return {'error': _("Could not extract company domain. Impossible to enrich the company.")}
+        company_data = self.env['res.partner'].enrich_company(company_domain, False, self.vat)
+        if company_data.get('error'):
+            return {'error': company_data.get('error_message', _('Unkown reason'))}
+
+        additional_data = company_data.pop('additional_info', False)
+
+        if company_data.get('state_id'):
+            company_data['state_id'] = company_data.get('state_id', {}).get('id', False)
+        if company_data.get('country_id'):
+            company_data['country_id'] = company_data.get('country_id', {}).get('id', False)
+
+        # Keep only not null value and value not already set on the target partner.
+        company_data = {field: value for field, value in company_data.items()
+                        if field in self.partner_id._fields and value and not self.partner_id[field]}
+
+        self.partner_id.write(company_data)
+        if additional_data:
+            template_values = json.loads(additional_data)
+            template_values['flavor_text'] = _("Company auto-completed by Odoo Partner Autocomplete Service")
+            self.partner_id.message_post_with_view(
+                'iap_mail.enrich_company',
+                values=template_values,
+                subtype_id=self.env.ref('mail.mt_note').id,
+            )
+        return {'error': False}
+
+    def _get_company_domain(self):
```

Comment:
Also could use a bit of docstring.

```py
""" Extract the company domain to be used by IAP services.
The domain is extracted from the website or the email information.
e.g: 
    - www.info.proximus.be -> proximus.be
    - info@proximus.be -> proximus.be """
```

And you will hate me but this could be easily unit tested :angel: